File: packages/ts/client/src/txbuilder/closePositionWithLiquidityOrca.ts
import {
  fetchAllTickArray,
  fetchMaybeWhirlpool,
  fetchPosition,
  getPositionAddress,
  getTickArrayAddress,
  WHIRLPOOL_PROGRAM_ADDRESS,
} from "@orca-so/whirlpools-client";
import { collectRewardsQuote, getTickArrayStartTickIndex, getTickIndexInArray } from "@orca-so/whirlpools-core";
import {
  Address,
  assertAccountsExist,
  GetAccountInfoApi,
  GetMultipleAccountsApi,
  IInstruction,
  Rpc,
  TransactionSigner,
} from "@solana/kit";
import { fetchAllMaybeMint, findAssociatedTokenPda, TOKEN_2022_PROGRAM_ADDRESS } from "@solana-program/token-2022";
import assert from "assert";

import {
  DEFAULT_ADDRESS,
  fetchAllVault,
  fetchMaybeTunaPosition,
  getClosePositionOrcaInstruction,
  getCreateAtaInstructions,
  getLendingVaultAddress,
  getTunaPositionAddress,
  HUNDRED_PERCENT,
  removeLiquidityOrcaInstruction,
} from "../index.ts";

export type ClosePositionWithLiquidityOrcaInstructionArgs = {
  swapToToken: number;
  minRemovedAmountA: number | bigint;
  minRemovedAmountB: number | bigint;
  maxSwapSlippage: number;
};

export async function closePositionWithLiquidityOrcaInstructions(
  rpc: Rpc<GetAccountInfoApi & GetMultipleAccountsApi>,
  authority: TransactionSigner,
  positionMint: Address,
  args: ClosePositionWithLiquidityOrcaInstructionArgs,
): Promise<IInstruction[]> {
  const tunaPosition = await fetchMaybeTunaPosition(rpc, (await getTunaPositionAddress(positionMint))[0]);
  if (!tunaPosition.exists) throw new Error("Tuna position account not found");

  const orcaPosition = await fetchPosition(rpc, (await getPositionAddress(positionMint))[0]);

  const whirlpool = await fetchMaybeWhirlpool(rpc, tunaPosition.data.pool);
  if (!whirlpool.exists) throw new Error("Whirlpool account not found");

  const [vaultA, vaultB] = await fetchAllVault(rpc, [
    (await getLendingVaultAddress(whirlpool.data.tokenMintA))[0],
    (await getLendingVaultAddress(whirlpool.data.tokenMintB))[0],
  ]);

  const [mintA, mintB, ...rewardMints] = await fetchAllMaybeMint(rpc, [
    whirlpool.data.tokenMintA,
    whirlpool.data.tokenMintB,
    ...whirlpool.data.rewardInfos.map(x => x.mint).filter(x => x !== DEFAULT_ADDRESS),
  ]);
  const allMints = [mintA, mintB, ...rewardMints];

  assert(mintA.exists, "Token A account not found");
  assert(mintB.exists, "Token B account not found");
  assertAccountsExist(rewardMints);

  const lowerTickArrayStartIndex = getTickArrayStartTickIndex(
    tunaPosition.data.tickLowerIndex,
    whirlpool.data.tickSpacing,
  );
  const [lowerTickArrayAddress] = await getTickArrayAddress(whirlpool.address, lowerTickArrayStartIndex);

  const upperTickArrayStartIndex = getTickArrayStartTickIndex(
    tunaPosition.data.tickUpperIndex,
    whirlpool.data.tickSpacing,
  );
  const [upperTickArrayAddress] = await getTickArrayAddress(whirlpool.address, upperTickArrayStartIndex);

  const [lowerTickArray, upperTickArray] = await fetchAllTickArray(rpc, [lowerTickArrayAddress, upperTickArrayAddress]);

  const lowerTick =
    lowerTickArray.data.ticks[
      getTickIndexInArray(tunaPosition.data.tickLowerIndex, lowerTickArrayStartIndex, whirlpool.data.tickSpacing)
    ];
  const upperTick =
    upperTickArray.data.ticks[
      getTickIndexInArray(tunaPosition.data.tickUpperIndex, upperTickArrayStartIndex, whirlpool.data.tickSpacing)
    ];

  const tunaPositionAta = (
    await findAssociatedTokenPda({
      owner: tunaPosition.address,
      mint: positionMint,
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    })
  )[0];

  const tunaPositionAtaA = (
    await findAssociatedTokenPda({
      owner: tunaPosition.address,
      mint: mintA.address,
      tokenProgram: mintA.programAddress,
    })
  )[0];

  const tunaPositionAtaB = (
    await findAssociatedTokenPda({
      owner: tunaPosition.address,
      mint: mintB.address,
      tokenProgram: mintB.programAddress,
    })
  )[0];

  //
  // Collect the list of instructions.
  //

  const instructions: IInstruction[] = [];
  const cleanupInstructions: IInstruction[] = [];

  //
  // Add token account creation instructions for every mint if needed.
  //

  const rewardsToClaim: number[] = [];
  const requiredMints: Address[] = [];
  requiredMints.push(whirlpool.data.tokenMintA);
  requiredMints.push(whirlpool.data.tokenMintB);

  let currentUnixTimestamp = BigInt(Math.floor(Date.now() / 1000));
  // It may happen in tests or in case of a wrong date.
  if (currentUnixTimestamp < whirlpool.data.rewardLastUpdatedTimestamp)
    currentUnixTimestamp = whirlpool.data.rewardLastUpdatedTimestamp;

  const rewardsQuote = collectRewardsQuote(
    whirlpool.data,
    orcaPosition.data,
    lowerTick,
    upperTick,
    currentUnixTimestamp,
  );

  for (let i = 0; i < rewardsQuote.rewards.length; i++) {
    if (rewardsQuote.rewards[i].rewardsOwed > 0n) {
      requiredMints.push(whirlpool.data.rewardInfos[i].mint);
      rewardsToClaim.push(i);
    }
  }

  for (const mintAddress of requiredMints) {
    const mint = allMints.find(mint => mint.address == mintAddress);
    assert(mint && mint.exists);

    const ixs = await getCreateAtaInstructions(authority, mint.address, authority.address, mint.programAddress);

    instructions.push(...ixs.init);
    cleanupInstructions.push(...ixs.cleanup);
  }

  // Add liquidity decrease instruction.
  instructions.push(
    await removeLiquidityOrcaInstruction(
      authority,
      tunaPosition,
      mintA,
      mintB,
      vaultA,
      vaultB,
      whirlpool,
      rewardsToClaim,
      rewardMints,
      {
        ...args,
        withdrawPercent: HUNDRED_PERCENT,
      },
    ),
  );

  // Close WSOL accounts if needed.
  instructions.push(...cleanupInstructions);

  // Add close position instruction.
  instructions.push(
    getClosePositionOrcaInstruction({
      mintA: mintA.address,
      mintB: mintB.address,
      authority,
      tunaPositionMint: positionMint,
      tunaPositionAta,
      tunaPositionAtaA,
      tunaPositionAtaB,
      orcaPosition: orcaPosition.address,
      tunaPosition: tunaPosition.address,
      whirlpoolProgram: WHIRLPOOL_PROGRAM_ADDRESS,
      tokenProgramA: mintA.programAddress,
      tokenProgramB: mintB.programAddress,
      token2022Program: TOKEN_2022_PROGRAM_ADDRESS,
    }),
  );

  return instructions;
}

File: rust-sdk/client/src/txbuilder/close_position_with_liquidity_orca.rs
use crate::accounts::{fetch_all_vault, fetch_tuna_position};
use crate::utils::get_create_ata_instructions;
use crate::{close_position_orca_instruction, get_tuna_position_address, get_vault_address, remove_liquidity_orca_instruction, RemoveLiquidityOrcaArgs, HUNDRED_PERCENT};
use anyhow::anyhow;
use orca_whirlpools_client::fetch_whirlpool;
use solana_client::rpc_client::RpcClient;
use solana_program::instruction::Instruction;
use solana_program::pubkey::Pubkey;

pub struct ClosePositionWithLiquidityOrcaArgs {
    pub swap_to_token: u8,
    pub min_removed_amount_a: u64,
    pub min_removed_amount_b: u64,
    pub max_swap_slippage: u32,
}

// TODO: rewards support
pub fn close_position_with_liquidity_orca_instructions(
    rpc: &RpcClient,
    authority: &Pubkey,
    position_mint: &Pubkey,
    args: ClosePositionWithLiquidityOrcaArgs,
) -> anyhow::Result<Vec<Instruction>> {
    let tuna_position = fetch_tuna_position(&rpc, &get_tuna_position_address(&position_mint).0)?;

    let whirlpool = fetch_whirlpool(rpc, &tuna_position.data.pool)?;
    let mint_a_address = whirlpool.data.token_mint_a;
    let mint_b_address = whirlpool.data.token_mint_b;

    let vaults = fetch_all_vault(&rpc, &[get_vault_address(&mint_a_address).0, get_vault_address(&mint_b_address).0])?;
    let (vault_a, vault_b) = (&vaults[0], &vaults[1]);

    let mut all_mint_addresses = vec![mint_a_address, mint_b_address];
    for reward_info in &whirlpool.data.reward_infos {
        if reward_info.mint != Pubkey::default() {
            all_mint_addresses.push(reward_info.mint);
        }
    }

    let mint_accounts = rpc.get_multiple_accounts(all_mint_addresses[0..all_mint_addresses.len()].into())?;
    let mint_a_account = mint_accounts[0].as_ref().ok_or(anyhow!("Token A mint account not found"))?;
    let mint_b_account = mint_accounts[1].as_ref().ok_or(anyhow!("Token B mint account not found"))?;

    let authority_ata_a_instructions = get_create_ata_instructions(&mint_a_address, authority, authority, &mint_a_account.owner, 0);
    let authority_ata_b_instructions = get_create_ata_instructions(&mint_b_address, authority, authority, &mint_b_account.owner, 0);

    let mut instructions = vec![];
    instructions.extend(authority_ata_a_instructions.create);
    instructions.extend(authority_ata_b_instructions.create);

    instructions.push(remove_liquidity_orca_instruction(
        authority,
        &tuna_position.data,
        &vault_a.data,
        &vault_b.data,
        &whirlpool.data,
        &mint_a_account.owner,
        &mint_b_account.owner,
        RemoveLiquidityOrcaArgs {
            withdraw_percent: HUNDRED_PERCENT,
            swap_to_token: args.swap_to_token,
            min_removed_amount_a: args.min_removed_amount_a,
            min_removed_amount_b: args.min_removed_amount_b,
            max_swap_slippage: args.max_swap_slippage,
        },
    ));

    instructions.extend(authority_ata_a_instructions.cleanup);
    instructions.extend(authority_ata_b_instructions.cleanup);

    instructions.push(close_position_orca_instruction(&authority, &tuna_position.data, &mint_a_account.owner, &mint_b_account.owner));

    Ok(instructions)
}
